### PR TITLE
Makes storage-engine pluggable 

### PR DESCRIPTION
**Pull request by simerplaha**
This PR is a WIP and not fully tested yet. Just looking for early feedback.

Makes core projects (`protocol`, `flow` etc) storage-engine independent. So storage-engines are pluggable now and there is no hard dependancy on RocksDB. SwayDB support is added to show how this work.

## How to change the storage engine?

Setting the value of `storageInThisBuild` in [build.sbt](https://github.com/alephium/alephium/compare/master...simerplaha:master#diff-5634c415cd8c8504fdb973a3ed092300b43c4b8fc1e184f7249eb29a55511f91R112) sets the target storage-engine.
```scala
// to enable RocksDB
lazy val storageInThisBuild = `storage-rocksdb`
//or for SwayDB
lazy val storageInThisBuild = `storage-swaydb`
```  

## `io` project split
I've had to split the `io` into smaller projects. 
- `io` - Contains only `IOResult` effect related code.
- `cache` - Cache & merkel-trees
- `storage` - Storage API implemented by external storage-engine libraries we want to support. Currently `storage-rocksdb` & `storage-swaydb`

## How to create and use storage-engine?
Note there are no RocksDB or SwayDB specific types here. 

```scala
//create the storage-engine
val storage: KeyValueSource =
  StorageInitializer.openUnsafe(
    path = dbPath, 
    setting = StorageSetting.syncWriteHDD(), 
    columns = List(ColumnFamily.Block)
  )

//use the above storage-engine to create a column for the Block column-family
val storageColumn: KeyValueStorage[String, String] = 
  KeyValueStorage[String, String](storage, ColumnFamily.Block)

//test
storageColumn.put("my key", "my value").isRight is true
storageColumn.get("my key") isE "my value"
```

Naming convention for `KeyValueSource` & `KeyValueStorage` is the same as before. Git history should be intact. 

Would renaming them to `StorageEngine` & `StorageColumn` be clearer?

### CODE REVIEW

**Comment by polarker** [comment]
@simerplaha it seems the styling check failed. you could run `make format` to fix it

**Comment by polarker** [comment]
All of the merkel trie related code are moved to `cache`, but only a few of them are using cache actually

File: flow/src/main/scala/org/alephium/flow/io/PendingTxStorage.scala [hunk 1]
```diff
@@ -21,9 +21,12 @@ import org.rocksdb.{ReadOptions, WriteOptions}
...
 import org.alephium.protocol.model.TransactionTemplate
-import org.alephium.storage.KeyValueStorage
-import org.alephium.storage.rocksdb.{RocksDBKeyValueCompanion, RocksDBKeyValueStorage, RocksDBSource}
-import org.alephium.storage.rocksdb.RocksDBSource.ColumnFamily
+import org.alephium.storage.{ColumnFamily, KeyValueStorage}
+import org.alephium.storage.rocksdb.{
+  RocksDBKeyValueCompanion,
+  RocksDBKeyValueStorage,
+  RocksDBSource
+}
```
Comment (polarker) [inline flow/src/main/scala/org/alephium/flow/io/PendingTxStorage.scala]: multi-line imports are discouraged

Comment (simerplaha) [reply flow/src/main/scala/org/alephium/flow/io/PendingTxStorage.scala]: This looks like an older commit. Should be single in the latest.

**Pull request closed without merging**